src/utils/SocketClient.hxx

```diff
@@ -254,34 +259,100 @@ private:
+                bool addr_okay = true;
+                if (addr_->ai_addr->sa_family != AF_INET)
                 {
+                    /* we only support IPv4 addresses */
+                    addr_okay = false;
+                }
+                if (addr_okay && disallowLocal_)
+                {
+                    /* We don't permit local connections to self, get the
+                     * addresses of all of our interfaces and check for this
+                     */
+                    struct ifaddrs *ifa;
+                    int result = getifaddrs(&ifa);
+                    if (result == 0)
                     {
-                        struct timeval tm;
-                        tm.tv_sec = timeoutSeconds_;
-                        tm.tv_usec = 0;
-                        ERRNOCHECK("setsockopt_timeout", setsockopt(fd_,
-                                                                    SOL_SOCKET,
-                                                                    SO_RCVTIMEO,
-                                                                    &tm,
-                                                                    sizeof(tm))
-                                  );
+                        bool has_address = false;
+                        struct ifaddrs *ifa_free = ifa;
+                        while (ifa)
+                        {
+                            if (ifa->ifa_addr)
+                            {
+                                /* ifa_addr pointer valid */
+                                if (ifa->ifa_addr->sa_family == AF_INET)
+                                {
+                                    /* have a valid IPv4 address */
+                                    struct sockaddr_in *ai_addr_in =
+                                        (struct sockaddr_in*)addr_->ai_addr;
+                                    struct sockaddr_in *if_addr_in =
+                                        (struct sockaddr_in*)ifa->ifa_addr;
+                                    if (ai_addr_in->sin_addr.s_addr ==
+                                        if_addr_in->sin_addr.s_addr)
+                                    {
+                                        /* trying to connected to myself */
+                                        addr_okay = false;
+                                    }
+                                    if (if_addr_in->sin_addr.s_addr != 0)
+                                    {
+                                        /* no IPv4 address is assigned */
+                                        has_address = true;
+                                    }
+                                }
+                            }
+                            ifa = ifa->ifa_next;
+                        }
+                        freeifaddrs(ifa_free);
+                        if (!has_address)
+                        {
+                            /* none of the network interfaces have a valid
```

Comment:
I think if we avoid checking has_address we should be fine as well.
The reason I suggest that is that having an implementation for getifaddr that just sets to nullptr is valid, without giving you protection.
Whereas has_address check will prevent all connections from happening with a really hard to debug error.
So if a new developer comes and implements a network driver for a new chip, they will be really puzzled.

Reply:
Perhaps but it would not be portable to Linux and MacOS.

Reply:
I don't get what would be the compatibility problem.
Just replace line 313 with ```addr_okay = true;```
This will work just fine on linux and mac. There is always a loopback interface anyway so you can never get into this codepath on linux and mac.